web3.js/src/connection.ts

```diff
@@ -4263,75 +4207,273 @@ export class Connection {
+                    args,
+                    e.message,
+                  );
+                }
+                // TODO: Maybe add an 'errored' state or a retry limit?
+                this._subscriptionsByHash[hash] = {
+                  ...subscription,
+                  state: 'pending',
+                };
+              } finally {
+                await this._updateSubscriptions();
+              }
+            })();
+            break;
+          case 'subscribed':
+            if (subscription.callbacks.size === 0) {
+              // By the time we successfully set up a subscription
+              // with the server, the client stopped caring about it.
+              // Tear it down now.
+              await (async () => {
+                const {serverSubscriptionId, unsubscribeMethod} = subscription;
+                this._subscriptionsByHash[hash] = {
+                  ...subscription,
+                  state: 'unsubscribing',
+                };
+                try {
+                  await this._rpcWebSocket.call(unsubscribeMethod, [
+                    serverSubscriptionId,
+                  ]);
+                } catch (e) {
+                  if (e instanceof Error) {
+                    console.error(`${unsubscribeMethod} error:`, e.message);
+                  }
+                  // TODO: Maybe add an 'errored' state or a retry limit?
+                  this._subscriptionsByHash[hash] = {
+                    ...subscription,
+                    state: 'subscribed',
+                  };
+                  await this._updateSubscriptions();
+                  return;
+                }
+                this._subscriptionsByHash[hash] = {
+                  ...subscription,
+                  state: 'unsubscribed',
+                };
+                await this._updateSubscriptions();
+              })();
+            }
+            break;
+          case 'subscribing':
+          case 'unsubscribing':
+            break;
+        }
+      }),
+    );
+  }
 
-    for (let clientSubscriptionId of rootClientSubscriptionIds) {
-      const sub = this._rootSubscriptions[clientSubscriptionId];
-      this._subscribe(sub, 'rootSubscribe', []);
+  /**
+   *
+   * @internal
+   */
+  _publishNotification<TCallback extends SubscriptionConfig['callback']>(
+    serverSubscriptionId: ServerSubscriptionId,
+    publishArgs: Parameters<TCallback>,
+  ): void {
+    const callbacks =
+      this._subscriptionCallbacksByServerSubscriptionId[serverSubscriptionId];
+    if (callbacks == null) {
+      return;
     }
-
-    for (let clientSubscriptionId of logsClientSubscriptionIds) {
-      const sub = this._logsSubscriptions[clientSubscriptionId];
-      let filter;
-      if (typeof sub.filter === 'object') {
-        filter = {mentions: [sub.filter.toString()]};
-      } else {
-        filter = sub.filter;
+    callbacks.forEach(cb => {
+      try {
+        cb(...(publishArgs as [any, any]));
```

Comment:
Should this be `any[]`? Some callbacks only have one argument, and future callbacks could have more than 2

Reply:
I had that initially, but TypeScript squaked: 

```
A spread argument must either have a tuple type or be passed to a rest parameter. ts(2556)
```